modules/caddypki/acmeserver/acmeserver.go

```diff
@@ -227,6 +259,62 @@ func (ash Handler) openDatabase() (*db.AuthDB, error) {
 	return database.(databaseCloser).DB, err
 }
 
+// makeClient creates an ACME client which will use a custom
+// resolver instead of net.DefaultResolver.
+func (ash Handler) makeClient() (acme.Client, error) {
+	for _, v := range ash.Resolvers {
+		addr, err := caddy.ParseNetworkAddress(v)
+		if err != nil {
+			// If a port wasn't specified for the resolver,
+			// try defaulting to 53 and parse again
+			if strings.Contains(err.Error(), "missing port in address") {
+				addr, err = caddy.ParseNetworkAddress(v + ":53")
+			}
+			if err != nil {
+				return nil, err
+			}
+		}
+		if addr.PortRangeSize() != 1 {
+			return nil, fmt.Errorf("resolver address must have exactly one address; cannot call %v", addr)
+		}
+		ash.resolvers = append(ash.resolvers, addr)
+	}
+
+	var resolver *net.Resolver
+	if len(ash.resolvers) != 0 {
+		dialer := &net.Dialer{
+			Timeout: 2 * time.Second,
+		}
+		resolver = &net.Resolver{
+			PreferGo: true,
+			Dial: func(ctx context.Context, network, address string) (net.Conn, error) {
+				//nolint:gosec
+				addr := ash.resolvers[weakrand.Intn(len(ash.resolvers))]
+				return dialer.DialContext(ctx, addr.Network, addr.JoinHostPort(0))
```

Comment:
I think this should use the requested `address` instead of `addr.Network` instead, since DNS can use both TCP and UDP.
The way it is now, the user has to specify the resolver address as `udp/127.0.0.1` to get it to work (at all).
```suggestion
				return dialer.DialContext(ctx, network, addr.JoinHostPort(0))
```

Reply:
Hmm, well it is intended to be able to specify the network in the address to override. But it is true that it currently defaults to `tcp` which is not ideal for a DNS resolver. I think I'll make another `ParseNetworkAddress` fn that allows specifying the default network, so it can default to `udp` instead.